substrate/primitives/blockchain/src/backend.rs

```diff
@@ -229,10 +228,37 @@ pub trait Backend<Block: BlockT>:
 	) -> std::result::Result<DisplacedLeavesAfterFinalization<Block>, Error> {
 		let mut result = DisplacedLeavesAfterFinalization::default();
 
-		if finalized_block_number == Zero::zero() {
+		let leaves = self.leaves()?;
+
+		// If we have only one leaf there are no forks, and we can return early.
+		if finalized_block_number == Zero::zero() || leaves.len() == 1 {
 			return Ok(result)
 		}
 
+		let first_leaf = leaves[0];
+		let leaf_block_header = self.expect_header(first_leaf)?;
+
+		// If the distance between the leafs and the finalized block is large,  calculating
+		// tree routes can be very expensive. In that case, we will try to find the
+		// lowest common ancestor between all the leaves. The assumption here is that the forks are
+		// close to the tip and not long. So the LCA can be computed from the header cache. If the
+		// LCA is above the finalized block, we know that there are no displaced leaves by the
+		// finalization.
+		if leaf_block_header
+			.number()
+			.checked_sub(&finalized_block_number)
+			.unwrap_or(0u32.into()) >
+			header_metadata::LRU_CACHE_SIZE.into()
```

Comment:
You should use the value for the forced canonicalization here.

Reply:
Why? I am not so sure what force canonicalization has to do with this? Maybe I am missing something, but afaik it only  affects for the state-db.

The idea with the cache size was that the tree route gets expensive once we read from disk. And the cache size is lower than the force canonicalization delay, wo we would benefit earlier.